File: swiftperl/swiftperl.c
//
//  swiftperl.c
//  swiftperl
//
//  Created by Dan Kogai on 7/12/14.
//  Copyright (c) 2014 Dan Kogai. All rights reserved.
//
#include <EXTERN.h>
#include <perl.h>
EXTERN_C void xs_init (pTHX);
EXTERN_C void boot_DynaLoader (pTHX_ CV* cv);
EXTERN_C void
xs_init(pTHX)
{
    char *file = __FILE__;
    dXSUB_SYS;
    
    /* DynaLoader is a special case */
    newXS("DynaLoader::boot_DynaLoader", boot_DynaLoader, file);
}

static PerlInterpreter *my_perl = NULL;

void swiftperl_sys_init() {
    int    noc = 0;
    char **nov = NULL;
    PERL_SYS_INIT3(&noc, &nov, &nov);
}
void *swiftperl_init () {
    if (my_perl != NULL) {
        fprintf(stderr, "%s\n",
                "Multiple perl instance is not supported yet\n");
        exit(-1);
    }
    char *embedding[] = { "", "-e", "1" };
    my_perl = perl_alloc();
    perl_construct(my_perl);
    perl_parse(my_perl, xs_init, 3, embedding, NULL);
    PL_exit_flags |= PERL_EXIT_DESTRUCT_END;
    perl_run(my_perl);
    return (void *)my_perl;
}
void swiftperl_deinit() {
    perl_destruct(my_perl);
    perl_free(my_perl);
    my_perl = NULL;
}
void swiftperl_sys_term() {
    if (my_perl != NULL) swiftperl_deinit();
    PERL_SYS_TERM();
}
void *swiftperl_eval_pv(const char* script, I32 croak_on_error) {
    return (void *)eval_pv(script, croak_on_error);
}
int swiftperl_err() {
    return SvTRUE(ERRSV);
}
char *swiftperl_errstr() {
    return SvPVx_nolen(ERRSV);
}
void *swiftperl_get_sv(char *name, int add) {
    return (void *)get_sv(name, add ? GV_ADD : 0);
}
// sv getters
int swiftperl_svok(void *vp) {
    return SvOK((SV *)vp);
}
int swiftperl_svtrue(void *vp) {
    return SvTRUE((SV *)vp);
}
unsigned long swiftperl_svuv(void *vp) {
    return (unsigned long)SvUV((SV *)vp);
}
long swiftperl_sviv(void *vp) {
    return (long)SvIV((SV *)vp);
}
double swiftperl_svnv(void *vp) {
    return (double)SvNV((SV *)vp);
}
char *swiftperl_svpv(void *vp) {
    return SvPV_nolen((SV *)vp);
}
// sv setters
void swiftperl_undef(void *vp) {
    sv_setsv_mg((SV *)vp, &PL_sv_undef);
}
void swiftperl_setuv(void *vp, unsigned long uv) {
    sv_setuv_mg((SV *)vp, uv);
}
void swiftperl_setiv(void *vp, long iv) {
    sv_setiv_mg((SV *)vp, iv);
}
void swiftperl_setnv(void *vp, double nv) {
    sv_setnv_mg((SV *)vp, nv);
}
void swiftperl_setpv(void *vp, char *pv) {
    sv_setpv_mg((SV *)vp, pv);
}
// av getters
void *swiftperl_get_av(char *name, int add) {
    return (void *)get_av(name, add ? GV_ADD : 0);
}
int swiftperl_av_len(void *vp){
    return av_len((AV *)vp);
}
void *swiftperl_av_fetch(void *vp, int key) {
    SV **svp = av_fetch((AV *)vp, key, 0);
    return svp ? (void *)*svp : NULL;
}
// hv getters
void *swiftperl_get_hv(char *name, int add) {
    return (void *)get_hv(name, add ? GV_ADD : 0);
}
int swiftperl_hv_iterinit(void *vp) {
    return hv_iterinit((HV *)vp);
}
void *swiftperl_hv_iternext(void *vp) {
    return (void *)hv_iternext((HV *)vp);
}
char *swiftperl_hv_iterkey(void *vp) {
    SV *ksv = hv_iterkeysv((HE *)vp);
    return SvPV_nolen(ksv);
}
void *swiftperl_hv_iterval(void *vp) {
    return HeVAL((HE *)vp);
}

File: swiftperl/swiftperl.swift
//
//  swiftperl.swift
//  swiftperl
//
//  Created by Dan Kogai on 7/12/14.
//  Copyright (c) 2014 Dan Kogai. All rights reserved.
//

class PerlSV : Printable {
    let sv:UnsafePointer<()>
    init (_ sv:UnsafePointer<()>){ self.sv = sv }
    var defined:Bool    { return swiftperl_svok(sv) != 0 }
    func undef() { swiftperl_undef(sv) }
    var asBool:Bool     {
    get {return swiftperl_svtrue(sv) != 0 }
    set { swiftperl_setiv(sv, newValue ? 1 : 0 ) }
    }
    var asUInt:UInt     {
    get { return swiftperl_svuv(sv) }
    set { return swiftperl_setuv(sv, newValue) }
    }
    var asInt:Int       {
    get { return swiftperl_sviv(sv) }
    set { swiftperl_setiv(sv, newValue) }
    }
    var asDouble:Double {
    get { return swiftperl_svnv(sv) }
    set { swiftperl_setnv(sv, newValue) }
    }
    var asString:String {
    get {
        return String.fromCString(CString(swiftperl_svpv(sv)))!
    }
    set {
        newValue.withCString {
            swiftperl_setpv(self.sv, $0)
        }
    }
    }
    var description:String { return asString }
    // References
    var refType: String {
        return String.fromCString(
            CString(swiftperl_reftype(sv))
        )!
    }
    func derefScalar() -> PerlSV? {
        return refType == "SCALAR"
            ? PerlSV(swiftperl_deref(sv)) : nil
    }
    func derefArray() -> PerlAV? {
        return refType == "ARRAY"
            ? PerlAV(swiftperl_deref(sv)) : nil
    }
    func derefHash() -> PerlHV? {
        return refType == "HASH"
            ? PerlHV(swiftperl_deref(sv)) : nil
    }
}

class PerlAV : Printable {
    let av:UnsafePointer<()>
    init (_ sv:UnsafePointer<()>){ self.av = sv }
    func toArray()->[PerlSV?] {
        var result = [PerlSV?]()
        for i in 0...swiftperl_av_len(av) {
            let sv = swiftperl_av_fetch(av, i, 0)
            result.append(sv == nil ? nil : PerlSV(sv))
        }
        return result
    }
    func get(i:Int)->PerlSV? {
        // it is NOT optional because it is always created
        let sv = swiftperl_av_fetch(av, Int32(i), 0)
        return sv == nil ? nil : PerlSV(sv)
    }
    func delete(i:Int)->PerlSV? {
        // it is NOT optional because it is always created
        let sv = swiftperl_av_delete(av, Int32(i))
        return sv == nil ? nil : PerlSV(sv)
    }
    subscript(i:Int)->PerlSV {
        // it is NOT optional because it is always created
        return PerlSV(swiftperl_av_fetch(av, Int32(i), 1))
    }
    var description:String { return "\(self.toArray())" }
}

class PerlHV : Printable {
    let hv:UnsafePointer<()>
    init (_ hv:UnsafePointer<()>){ self.hv = hv }
    func toDictionary()->[String:PerlSV?] {
        var result = [String:PerlSV?]()
        swiftperl_hv_iterinit(hv)
        while true {
            let he = swiftperl_hv_iternext(hv)
            if he == nil { break }
            let key = String.fromCString(
                CString(swiftperl_hv_iterkey(he))
                )!
            result[key] = PerlSV(swiftperl_hv_iterval(he))
        }
        return result
    }
    func get(k:String)->PerlSV? {
        let sv = swiftperl_hv_fetchs(hv, k.withCString{$0}, 0)
        return sv == nil ? nil : PerlSV(sv)
    }
    func delete(k:String)->PerlSV? {
        let sv = swiftperl_hv_delete(hv, k.withCString{$0})
        return sv == nil ? nil : PerlSV(sv)
    }
    subscript(k:String)->PerlSV {
        // it is NOT optional because it is always created
        return PerlSV(
            swiftperl_hv_fetchs(hv, k.withCString{$0}, 1)
            )
    }
    var description:String { return "\(self.toDictionary())" }
}

class Perl {
    var debug:Bool
    init(debug:Bool=false) {
        self.debug = debug
        swiftperl_init()
    }
    deinit  {
        if debug { println("Perl: deinitializing interpreter") }
        swiftperl_deinit()
    }
    func reinit() {
        if debug { println("Perl: reinitializing interpreter") }
        swiftperl_deinit(); swiftperl_init()
    }
    class func sysInit() { swiftperl_sys_init() }
    class func sysTerm() { swiftperl_sys_term() }
    var preamble = "use v5.16; no strict;"
    func eval(script:String)->PerlSV {
        return  (preamble+script).withCString {
            PerlSV(swiftperl_eval_pv($0, 0))
        }
    }
    var evalok:Bool {
        return swiftperl_err() == 0
    }
    var errstr:String {
        return String.fromCString(CString(swiftperl_errstr()))!
    }
    func use(name:String)->PerlSV {
        return self.eval("use \(name);")
    }
    func use<T>(name:String, _ args:T...)->PerlSV {
        return self.use(
            name + " " + ",".join(args.map{"'\($0)'"})
        )
    }
    func sv(name:String, _ add:Bool=false) -> PerlSV? {
        let sv = name.withCString {
            swiftperl_get_sv($0, add ? 1 : 0)
        }
        return sv == nil ? nil : PerlSV(sv)
    }
    func $(name:String, _ add:Bool=false)->PerlSV? {
        return self.sv(name, add)
    }
    func av(name:String, _ add:Bool=false)->PerlAV? {
        let av = name.withCString {
            swiftperl_get_av($0, add ? 1 : 0)
        }
        return av == nil ? nil : PerlAV(av)
    }
    func hv(name:String, _ add:Bool=false)->PerlHV? {
        let hv = name.withCString {
            swiftperl_get_hv($0, add ? 1 : 0)
        }
        return hv == nil ? nil : PerlHV(hv)
    }
}
